Improve accessibility on the front-end

edwilde opened the pull request:
We use better navigator across a bunch of projects. The main aim of this PR was to fix the accessibility warning which we get when running automated tools on local Silverstripe instances using BetterNavigator.

<img width="624" alt="Screen Shot 2023-07-31 at 9 55 14 AM" src="https://github.com/jonom/silverstripe-betternavigator/assets/415374/8b31c12e-363c-48ae-84ed-84aa91db6095">

However, it got a little more involved so here is a summary of changes 🚀 
* Replaced compass with a more modern toolset using sass and svg-sprite
* Replaced the 1x resolution icons with a compressed svg vector sprite so they are nice and crisp on high-definition screens
* Updated the colour contrast across various elements to meet WCAG 2.1 AA standard (4.5:1)
* Updated the markup to be more semantic and work nicely with the accessibility tree
* Added a couple of `aria-` tags to communicate state and relationship of the tool expansion
* Fixed a couple of alignment issues with the button labels

<img width="607" alt="Screen Shot 2023-07-31 at 4 56 31 PM" src="https://github.com/jonom/silverstripe-betternavigator/assets/415374/301dc116-4ccc-4e2b-a06b-ab1597b7673c">

I really hope this is useful and not too much of an overhead to look through.

edwilde committed:
Add some basic sass tooling to replace compass

edwilde committed:
Replace bitmap sprite with a vector sprite

and _roughly_ the same icons

edwilde committed:
Remove old compass config

edwilde committed:
Fix some visual accessibility issues

edwilde committed:
Add some accessibility fixes to the markup and script

edwilde committed:
Update demo image

With new visual styling

edwilde committed:
Remove unused dependencies

jonom commented:
Wow! This is great, thanks. The only push back I have is that I don't like the brown tint of the kelp green. How about the green used here on github (`rgb(31, 136, 61)`) would that meet contrast requirements? I'm also guessing that the default dev colour is even lower contrast, do you want to look at updating that too?

jonom commented:
nvm I see you did change the draft colour as well. I can't say I'm crazy about the new colors. How much larger would the text need to be for lower contrast to be acceptable?

jonom commented on templates/JonoM/BetterNavigator/BetterNavigator.ss:
```diff
@@ -43,98 +44,96 @@
         <% if $Mode == 'dev' || $IsDeveloper %>
 
-            <div class="bn-heading"><%t JonoM\BetterNavigator.DEVELOPER_TOOLS_HEADING 'Developer tools' %></div>
+        <h2 class="bn-heading"><%t JonoM\BetterNavigator.DEVELOPER_TOOLS_HEADING 'Developer tools' %></h2>
```
I know it's not semantic, but divs were used here to avoid picking up any default styles from the project. If you want to use semantic elements you might need to apply a reset to each.

edwilde replied:
ahh, ok, yes I hadn't thought of that.

I have added a few extra resets to various elements. Because everything is under an `id` selector, we have pretty high specificity so there actually weren't too many things to add.

jonom commented on images/bn-icon/tick.svg:
This and the close icon look a bit chunky, are there slimmer versions you can use? And do you need to add any license info to use these icons?

jonom commented on scss/_iconsprite.scss:
```diff
@@ -0,0 +1,114 @@
+%svg-common {
```
I certainly appreciate the effort to continue using a sprite map but I'm wondering about the maintenance burden. What do you think about loading the individual svg files directly instead? With http/2 any static assets like these often being served up by a proxy that supports it (e.g. CloudFlare), there probably isn't much of a gain in using a sprite map - especially since BetterNavigator isn't seen by the end users.

If you do remove the sprite feature, we might not even need sass. Might be nice to fully modernise the css by using css variables and css nesting, with perhaps [PostCSS](https://github.com/postcss/postcss-cli) or [Lightning CSS](https://lightningcss.dev/transpilation.html) to transpile for older browsers. But not a requirement.

jonom reviewed (changes requested):
Thanks @edwilde, I'll be very happy to have sharper icons and a more modern css workflow.

I'm not sold on the color change yet, so I'm interested to see what else we can do. Perhaps having a badge effect around the text could be an option: 
![image](https://github.com/jonom/silverstripe-betternavigator/assets/1079425/0885d10b-438e-480d-9f3d-83ffc58c22c9)

I'm also a bit skeptical about blindly following contrast scores, here's an example from [UXMovement](https://uxmovement.com/buttons/the-myths-of-color-contrast-accessibility/) that illustrates they're not perfect.
![image](https://github.com/jonom/silverstripe-betternavigator/assets/1079425/60dad925-03c8-42d7-ad2a-7539184453d3)

I wouldn't expect users to read a page of text with the combinations we have currently but maybe they're fine for their limited use, and the fact that the colour does most of the communicating without the text? I'm a bit reluctant to make changes to the design if the use case is to satisfy an automated testing suite, rather than responding to the feedback and needs of real users.

edwilde committed:
Add information on icon sprite generation and licenses

edwilde commented:
> How about the green used here on github (rgb(31, 136, 61)) would that meet contrast requirements?

Hah, interestingly Github doesn't meet WCAG 2.1 AA here (but only just).

<img width="579" alt="Screen Shot 2023-08-01 at 9 54 28 AM" src="https://github.com/jonom/silverstripe-betternavigator/assets/415374/0e86a7e5-466b-425d-a9f2-987f52c3a9f8">

edwilde commented:
> How much larger would the text need to be for lower contrast to be acceptable?

According to [WebAIM](https://webaim.org/articles/contrast/#:~:text=WCAG%20defines%20large%20text%20as,larger%20if%20it%20is%20bold.), actually not much bigger to get to the 3:1 contrast requirement.

> WCAG defines large text as text that is 18pt and larger, or 14pt and larger if it is bold.

edwilde committed:
Put the close icon on a diet

Replace with a less chunky one

edwilde committed:
Reduce the size of the tick icon

To match closer to the previous png icons

edwilde committed:
Fix issue with reloading the page when expanded

edwilde committed:
Brighten the status colours

Because of the increase in font size, we are now able to use a lower contrast of 3:1 to follow WCAG 2.1 AA. The draft colour reverts to the previous, the live colour is a very slight change.

edwilde committed:
Drop the size of the bug icon a little

edwilde committed:
Add resets for semantic tags used

edwilde committed:
Fix indentation

edwilde committed:
Update readme image with brighter colours

edwilde commented:
> I'm not sold on the color change yet, so I'm interested to see what else we can do.

Yes, certainly worth a second look. Turns out the status was using bold text at `14px` which _just_ meets the 3:1 contrast requirement instead of 4.5:1. So the colour changes could be much more minimal and keep a very similar brightness. 🎉

jonom reviewed (changes requested):
Sorry about the long delay here @edwilde, summer plans etc. got in the way. The colours and build process looks good, just some minor issues with styles to clean up if you don't mind.

edwilde committed:
Fix incorrect closing tag

Well spotted @jonom!

edwilde commented on templates/JonoM/BetterNavigator/BetterNavigator.ss:
```diff
@@ -17,22 +18,22 @@
                         <a href="$LiveLink.Link"><span class="bn-icon-view"></span><%t JonoM\BetterNavigator.VIEW_LIVE_LABEL 'View live' %></a>
                     <% else %>
-                        <span class="bn-disabled"><span class="bn-icon-view"></span><%t JonoM\BetterNavigator.NOT_YET_PUBLISHED_LABEL 'Not yet published' %></span>
+                        <a href="$LiveLink.Link" aria-disabled="true" class="bn-disabled"><span class="bn-icon-view"></a><%t JonoM\BetterNavigator.NOT_YET_PUBLISHED_LABEL 'Not yet published' %></a>
```
Ah, nards! 🤦

edwilde committed:
Update to use a single button trigger and avoid style inheritance

edwilde committed:
Add rebuilt demo image

edwilde commented on templates/JonoM/BetterNavigator/BetterNavigator.ss:
```diff
@@ -1,10 +1,11 @@
-    <div id="BetterNavigatorStatus" class="$Viewing">
-        <span class="bn-icon-cog"></span>
+    <h2 id="BetterNavigatorStatus" class="$Viewing" tabindex="-1">
+        <button id="BetterNavigatorStatusButton" aria-expanded="false" aria-controls="BetterNavigator" aria-label="<%t JonoM\BetterNavigator.ARIA_MENU_LABEL 'Tool menu' %>">
```
Yep, absolutely right...although this required a bit more of a styling change than I expected

edwilde commented:
> Turns out the status was using bold text at 14px which just meets the 3:1 contrast requirement instead of 4.5:1

Nope, I was wrong...it needed to be `18.5px` minimum to work with the 3:1 contrast requirement ([ref](https://www.w3.org/WAI/WCAG21/Understanding/contrast-minimum.html#:~:text=equivalent%20to%20approximately%2018.5px%20and%2024px.)).

Which means, and you might not like this, I had to increase the text size 👀  ... It was a bit jarring at first, but after a while I got used to it.

![demo](https://github.com/jonom/silverstripe-betternavigator/assets/415374/75c2cbda-225c-42e0-8c95-7f676b536dc8)

jonom commented:
> Which means, and you might not like this, I had to increase the text size 👀

Oof 😂 that is really big. Too big for me, sorry. I'm going to ask you to either increase that font-size with your own stylesheet, or make it a config opt-in option with something like `wcag: AA`. I'm just not convinced that this specific piece is a real-world accessibility issue.

jonom commented:
Out of curiosity - did you test the 'Developer Tools' and 'Debugging' sub-titles for contrast?

edwilde commented:
> did you test the 'Developer Tools' and 'Debugging' sub-titles for contrast?

Of course 👍 

<img width="269" alt="Screen Shot 2023-10-30 at 8 06 46 AM" src="https://github.com/jonom/silverstripe-betternavigator/assets/415374/d3c73f12-1f66-4a56-b303-96a1af1e8b3d">

edwilde commented:
To be honest, I have been working with this design for a while now and I think it is an improvement. The navigator is much clearer and easier to see at a glance on most pages and fits better with the larger text styles used on modern websites. When open it is easier to read, see the comparison below:

# Before

<img width="524" alt="Screen Shot 2023-11-08 at 8 46 39 AM" src="https://github.com/jonom/silverstripe-betternavigator/assets/415374/b1e9174e-ff8d-4d54-be05-fa2c87c0c8ac">

# After

<img width="531" alt="Screen Shot 2023-11-08 at 8 45 02 AM" src="https://github.com/jonom/silverstripe-betternavigator/assets/415374/651290db-cf09-45ec-8e20-5d8c299d0465">

# Before

<img width="242" alt="Screen Shot 2023-11-08 at 8 53 07 AM" src="https://github.com/jonom/silverstripe-betternavigator/assets/415374/0fbed066-abe7-4b57-845b-c4aab629d5ab">

# After

<img width="242" alt="Screen Shot 2023-11-08 at 8 40 39 AM" src="https://github.com/jonom/silverstripe-betternavigator/assets/415374/bdc48b61-5bb2-429c-b5d5-16186336df00">

I guess it is up to you, I don't believe in making accessibility an _opt-in feature_, as we can see above the changes have both aesthetic and usability benefits for _all_ users.

jonom commented:
I'm certainly not arguing that this module shouldn't be accessible, I'm just not sure that contrast alone should be used to determine if there's a genuine accessibility problem here. I understand your perspective though and do agree that we don't want to send a message that accessibility is optional (I was meaning more that satisfying strict contrast rules could be optional). The text feels way too big to me but I don't want to change to having a neutral background and black text with a smaller coloured area to allow for a smaller text size as I like having the big block of colour to reference. So I'll put my personal tastes aside and accept the giant text 😬. Maybe I'll come to love it. I'm going to merge now but I want to do a little more testing before I tag a release. Thanks again for all of your efforts, patience and persistence!

edwilde commented:
Thanks @jonom, I really hope you'll get to like it over time like I did 🤩

jonom commented:
I'm sorry, I couldn't handle it 😅 I have reduced the size and added a darker background behind to meet the AA contrast spec.

![image](https://github.com/jonom/silverstripe-betternavigator/assets/1079425/9f211880-8006-4f41-96c9-deb341f8ac28) ![image](https://github.com/jonom/silverstripe-betternavigator/assets/1079425/3d5988ab-c8c0-4bf3-95c8-cfc8919f0d3e)

edwilde commented:
Nice balance 👍
